Channel pools in vhost

BorePlusPlus opened the pull request:
To get around the bottleneck of creating one channel at a time for publishing,
I added channel pools in vhost (to be used by publications). This is my
inital work and still rough around the edges (pool size hardcoded,
not sure I like how code looks with proliferation of channel
retrieving/returning methods, etc.)
Putting it here as a conversation starter and I have intention to keep
working on it.

Tests still pass and ones testing publication of lots of messages are
appropriately faster. I did not add any new tests specific to channel
pooling.

I used 2.5 branch of generic-pool project as newer versions (3+) require es6
features.

Let me know what you think.

BorePlusPlus commented:
Just realised I should drain channels from the pool on pause. So when we restart, new ones will be created using new connection.

cressie176 commented:
Tests are much faster (especially the ones which publish multiple messages)

cressie176 commented:
> Just realised I should drain channels from the pool on pause. So when we restart, new ones will be created using new connection.

What about comparing the connection (or a token generated at connection time) in the pool's validate method? Then if the underlying connection had been replaced, the pool would automatically allocate new channels as needed

cressie176 commented on lib/amqp/Publication.js:
```diff
@@ -46,13 +46,17 @@ function Publication(getChannelFn, publishFn, vhost, config) {
-                emitter.emit('success', publishConfig.options.messageId)
+              returnChannelFn(vhost, channel, errorHandler, returnHandler)
+              if (err) return emitter.emit('error', err)
+              emitter.emit('success', publishConfig.options.messageId)
```
Indentation (should be 4 chars)

BorePlusPlus replied:
:disappointed: yep, specially bad as I am a 4 space aficionado

cressie176 commented on lib/amqp/Publication.js:
```diff
@@ -46,13 +46,17 @@ function Publication(getChannelFn, publishFn, vhost, config) {
-                if (err) return emitter.emit('error', err)  // Channel will already be closed, reclosing will trigger an error
-                channel.close()
-                emitter.emit('success', publishConfig.options.messageId)
+              returnChannelFn(vhost, channel, errorHandler, returnHandler)
```
Agree we need to clean up the channel listeners, but maybe we shouldn't return the channel to the pool on error?

BorePlusPlus replied:
We need to return channels to the pool, otherwise we'll starve them. 

Also on the topic of publish errors. If I understood amqplib code correctly, the only error this callback will get is this [nack](https://github.com/squaremo/amqp.node/blob/31bb2a9c40219ce2fc8c4fede15a84a9f54b36a1/lib/channel.js#L34). Which doesn't really signal that channel itself errored, so I think channel might actually be OK for reuse. And we also have a [handler for channel error events](https://github.com/BorePlusPlus/rascal/blob/c37fc5704e7ce228411091937ffb8cd7afaa2650/lib/amqp/Vhost.js#L114), which will mark channel as closed on error, so it won't get reused.

cressie176 replied:
kk

BorePlusPlus commented on lib/amqp/Publication.js:
```diff
@@ -46,13 +46,17 @@ function Publication(getChannelFn, publishFn, vhost, config) {
+        var returnHandler = emitter.emit.bind(emitter, 'return')
+        borrowChannelFn(vhost, errorHandler, returnHandler, function(err, channel) {
+            if (err) {
+                if (channel) returnChannelFn(vhost, channel, errorHandler, returnHandler)
```
Belt and braces. But by just looking at the functions we're calling here you can't ever get both. So I'll get rid of it.

BorePlusPlus commented on lib/amqp/Publication.js:
```diff
@@ -78,22 +82,36 @@ function Publication(getChannelFn, publishFn, vhost, config) {
+function returnConfirmChannel(vhost, channel, errorHandler, returnHandler) {
+    vhost.returnConfirmChannel(channel)
+    channel.removeListener('error', errorHandler)
+    channel.removeListener('return', returnHandler)
```
Much appreciated.

cressie176 replied:
There's a test for this, and it still passes, so should be ok.

BorePlusPlus commented on lib/amqp/Vhost.js:
```diff
@@ -87,13 +92,52 @@ function Vhost(config) {
+                    next(null, channel)
+
+                    function markClosed() {
+                        channel._closed = true
```
Was thinking of making it more unique, namespacing with `_rascal` sounds good.

BorePlusPlus commented:
> What about comparing the connection (or a token generated at connection time) in the pool's validate method? Then if the underlying connection had been replaced, the pool would automatically allocate new channels as needed

Excellent idea will look into doing that..

BorePlusPlus commented:
Things we talked about in the comments are addressed.

cressie176 commented on README.md:
```diff
@@ -193,6 +191,18 @@ If you specify an array of connections instead of a single connection object Ras
+{
+  "vhosts": {
+    "v1": {
+      "channelPoolSize": 3
```
Maybe we should have separate arguments for confirm and regular. It could get pretty wasteful having a pool of 10 channels which never got used.

BorePlusPlus replied:
Yep was thinking about it, but wasn't sure if added complexity is warranted, considering there are about 65k channel available, so I just stuck with one setting at the moment.

If we go for explicit two settings, would you just call them `channelPoolSize` and 'confirmChannelPoolSize`? Would you perhaps nest them under something like `publicationChannelPool`?

BorePlusPlus commented:
Updated as per your comments:
 - debug pool stats
 - split configuration for publication pools
 - add/remove channel listeners in separate function
 - use debug's own formatting

cressie176 commented:
Thanks @BorePlusPlus. Really appreciate all the hard work. A big improvement to Rascal.

cressie176 commented:
Published as rascal@1.3.0. Thanks again
